src/NetMQ/NetMQPoller.cs

```diff
@@ -259,15 +297,27 @@ public void Add([NotNull] Socket socket, [NotNull] Action<Socket> callback)
         /// <exception cref="ArgumentNullException">If socket is null</exception>
         /// <exception cref="ArgumentException">If socket is already disposed</exception>
         /// <exception cref="InvalidOperationException">If socket is getting disposed during the operation</exception>
+        [Obsolete("Queues the action on the poller's thread, but provides no sync mechanism.  Please use RemoveAsync() instead")]
         public void Remove(ISocketPollable socket)
+        {
+            // keeps the current public API intact, though flawed (no way to know when the task actually removes the socket).
+            RemoveAsync(socket);
+        }
+
+        public Task RemoveAsync(ISocketPollable socket)
         {
             if (socket == null)
                 throw new ArgumentNullException(nameof(socket));
+
+            // JASells: not sure I agree with this thow.
+            // If trying to remove a disposed socket, why complain?  It *might* get removed before the poller thread accesses it.  
+            // The issue is if the poller's thread tries to actually service the disposed socket before the remove call...
```

Comment:
I believe the intent here was to reliably fail when things aren't torn down in the correct order. IIRC, by not throwing here it was possible that code would only fail _some_ of the time, which (perhaps counterintuitively) is worse than failing all of the time.

Reply:
I belive the fix I made [here ](https://github.com/jasells/netmq/blob/9f17258c917a94422a87a7532aea3f29b3d0b58c/src/NetMQ/NetMQSelector.cs#L169)is 1/2 of the required fix to be able to alleviate this concern entirely.

Would need to do the same check [here](https://github.com/jasells/netmq/blob/9f17258c917a94422a87a7532aea3f29b3d0b58c/src/NetMQ/NetMQSelector.cs#L130).

Reply:
Sorry forgot I agreed to separate this...  I will do that ASAP, so as to keep things cleaner.

Reply:
Removed this comment from source.